quinn/src/lib.rs

```diff
@@ -798,93 +618,220 @@ impl Connection {
-            .connection(self.0.handle)
-            .rem_cid()
+        self.0.borrow().inner.rem_cid()
     }
 
     /// The negotiated application protocol
     pub fn protocol(&self) -> Option<Box<[u8]>> {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .protocol()
-            .map(|x| x.into())
+        self.0.borrow().inner.protocol().map(|x| x.into())
     }
 
     // Update traffic keys spontaneously for testing purposes.
     #[doc(hidden)]
     pub fn force_key_update(&self) {
-        self.0
-            .endpoint
-            .borrow_mut()
-            .inner
-            .force_key_update(self.0.handle)
+        self.0.borrow_mut().inner.force_key_update()
     }
 }
 
 struct ConnectionInner {
-    endpoint: Rc<RefCell<EndpointInner>>,
+    log: Logger,
+    inner: quinn::Connection,
+    driver: Option<Task>,
     handle: ConnectionHandle,
+    pending: Pending,
     side: Side,
+    // TODO: Replace this with something custom that avoids using oneshots to cancel
+    timers: FuturesUnordered<Timer>,
+    conn_events: mpsc::UnboundedReceiver<ConnectionEvent>,
+    endpoint_events: mpsc::UnboundedSender<(ConnectionHandle, EndpointEvent)>,
+    connected: bool,
+    closed: bool,
 }
 
-impl Drop for ConnectionInner {
-    fn drop(&mut self) {
-        let endpoint = &mut *self.endpoint.borrow_mut();
-        if let hash_map::Entry::Occupied(mut pending) = endpoint.pending.entry(self.handle) {
-            if pending.get().drained {
-                pending.remove();
-                return;
+impl ConnectionInner {
+    fn drive_transmit(&mut self, now: Instant) {
+        while let Some(t) = self.inner.poll_transmit(now) {
+            self.endpoint_events
+                .unbounded_send((self.handle, EndpointEvent::Transmit(t)))
+                .unwrap();
+        }
+    }
+
+    fn handle_endpoint_events(&mut self) {
+        while let Some(event) = self.inner.poll_endpoint_events() {
+            if let quinn::EndpointEvent::Closed { .. } = &event {
+                self.closed = true;
+                self.pending
+                    .fail(ConnectionError::TransportError(quinn::TransportError {
+                        code: quinn::TransportErrorCode::NO_ERROR,
+                        frame: None,
+                        reason: "connection is closing".to_string(),
+                    }));
+            }
+            self.endpoint_events
+                .unbounded_send((self.handle, EndpointEvent::Proto(event)))
+                .unwrap();
+        }
+    }
+
+    fn handle_conn_events(&mut self) -> Result<(), ()> {
+        loop {
+            match self.conn_events.poll() {
+                Ok(Async::Ready(Some(ConnectionEvent::Proto(event)))) => {
+                    self.inner.handle_event(event);
+                }
+                Ok(Async::Ready(Some(ConnectionEvent::Close(_)))) => {
+                    self.closed = true;
```

Comment:
We need to call `Pending::fail` here to ensure all downstream futures are notified.

Reply:
Done.